Compiler crashes with an invocation of a polymorphic by-name function in an if

## Compiler version
3.4.0-RC3

## Minimized code
```Scala
def byName[T](p: => T): T = p
val test = (if ??? then byName else (??? : ((=> Int) => Int))) (42)
```

## Output (click arrow to expand)
<details>

```scala
[debug] [zinc] IncrementalCompile -----------
[debug] IncrementalCompile.incrementalCompile
[debug] previous = Stamps for: 2 products, 1 sources, 1 libraries
[debug] current source = Set(${BASE}/src/main/scala/Main.scala)
[debug] > initialChanges = InitialChanges(Changes(added = Set(), removed = Set(), changed = Set(${BASE}/src/main/scala/Main.scala), unmodified = ...),Set(${BASE}/target/scala-3.4.0-RC3/classes/Main$package$.class, ${BASE}/target/scala-3.4.0-RC3/classes/Main$package.class),Set(),API Changes: Set())
[debug] 
[debug] Initial source changes:
[debug]         removed: Set()
[debug]         added: Set()
[debug]         modified: Set(${BASE}/src/main/scala/Main.scala)
[debug] Invalidated products: Set(${BASE}/target/scala-3.4.0-RC3/classes/Main$package$.class, ${BASE}/target/scala-3.4.0-RC3/classes/Main$package.class)
[debug] External API changes: API Changes: Set()
[debug] Modified binary dependencies: Set()
[debug] Initial directly invalidated classes: Set(Main$package)
[debug] Sources indirectly invalidated by:
[debug]         product: Set(${BASE}/src/main/scala/Main.scala)
[debug]         binary dep: Set()
[debug]         external source: Set()
[debug] all 1 sources are invalidated
[debug] Initial set of included nodes: Main$package
[debug] Recompiling all sources: number of invalidated sources > 50.0 percent of all sources
[debug] compilation cycle 1
[info] compiling 1 Scala source to /Users/chen/ChenData/Programming/Scala/Test/target/scala-3.4.0-RC3/classes ...
[debug] Returning already retrieved and compiled bridge: /Users/chen/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-sbt-bridge/3.4.0-RC3/scala3-sbt-bridge-3.4.0-RC3.jar.
[debug] [zinc] Running cached compiler 617b6cdd for Scala Compiler version 3.4.0-RC3
[debug] [zinc] The Scala compiler is invoked with:
[debug]         -bootclasspath
[debug]         /Users/chen/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.4.0-RC3/scala3-library_3-3.4.0-RC3.jar:/Users/chen/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar
[debug]         -classpath
[debug]         /Users/chen/ChenData/Programming/Scala/Test/target/scala-3.4.0-RC3/classes
[error] ## Exception when compiling 1 sources to /Users/chen/ChenData/Programming/Scala/Test/target/scala-3.4.0-RC3/classes
[error] java.lang.AssertionError: assertion failed: ExprType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Int)) in in intersection () ?=> Any & => Int
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.core.Types$.expectValueTypeOrWildcard(Types.scala:3651)
[error] dotty.tools.dotc.core.Types$AndType$.apply(Types.scala:3466)
[error] dotty.tools.dotc.core.Types$AndType$.balanced(Types.scala:3484)
[error] dotty.tools.dotc.core.TypeComparer.andType$$anonfun$1(TypeComparer.scala:2541)
[error] dotty.tools.dotc.core.TypeComparer.liftIfHK(TypeComparer.scala:2579)
[error] dotty.tools.dotc.core.TypeComparer.andTypeGen(TypeComparer.scala:2519)
[error] dotty.tools.dotc.core.TypeComparer.andType(TypeComparer.scala:2541)
[error] dotty.tools.dotc.core.TypeComparer.glb(TypeComparer.scala:2354)
[error] dotty.tools.dotc.core.TypeComparer.lubArgs(TypeComparer.scala:2413)
[error] dotty.tools.dotc.core.TypeComparer$.lubArgs(TypeComparer.scala:3169)
[error] dotty.tools.dotc.core.TypeOps$.mergeRefinedOrApplied$1(TypeOps.scala:272)
[error] dotty.tools.dotc.core.TypeOps$.orDominator(TypeOps.scala:393)
[error] dotty.tools.dotc.core.Types$OrType.join(Types.scala:3546)
[error] dotty.tools.dotc.core.Types$Type.goOr$1(Types.scala:916)
[error] dotty.tools.dotc.core.Types$Type.go$1(Types.scala:789)
[error] dotty.tools.dotc.core.Types$Type.findMember(Types.scala:932)
[error] dotty.tools.dotc.core.Types$Type.memberBasedOnFlags(Types.scala:726)
[error] dotty.tools.dotc.core.Types$Type.nonPrivateMember(Types.scala:716)
[error] dotty.tools.dotc.core.Types$NamedType.memberDenot(Types.scala:2508)
[error] dotty.tools.dotc.core.Types$NamedType.reload$1(Types.scala:2841)
[error] dotty.tools.dotc.core.Types$NamedType.withPrefix(Types.scala:2856)
[error] dotty.tools.dotc.core.Types$NamedType.derivedSelect(Types.scala:2789)
[error] dotty.tools.dotc.ast.tpd$TypedTreeCopier.Select(tpd.scala:640)
[error] dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:241)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:448)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:291)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:450)
[error] dotty.tools.dotc.transform.MegaPhase.mapValDef$1(MegaPhase.scala:247)
[error] dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:252)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:448)
[error] dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:461)
[error] dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:461)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:372)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:450)
[error] dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:268)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:448)
[error] dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:461)
[error] dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:461)
[error] dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:392)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:395)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:450)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:477)
[error] dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:489)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:354)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:360)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:315)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:337)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:348)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:357)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:71)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:357)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:261)
[error] dotty.tools.dotc.Run.compile(Run.scala:246)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:180)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:178)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:464)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:419)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:406)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:172)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2371)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2321)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:31)
[error] sbt.internal.io.Retry$.apply(Retry.scala:47)
[error] sbt.internal.io.Retry$.apply(Retry.scala:29)
[error] sbt.internal.io.Retry$.apply(Retry.scala:24)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2319)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:69)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[error] sbt.Execute.work(Execute.scala:292)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[error] java.base/java.lang.Thread.run(Thread.java:1589)
[error]            
[error] java.lang.AssertionError: assertion failed: ExprType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Int)) in in intersection () ?=> Any & => Int
[error]         at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error]         at dotty.tools.dotc.core.Types$.expectValueTypeOrWildcard(Types.scala:3651)
[error]         at dotty.tools.dotc.core.Types$AndType$.apply(Types.scala:3466)
[error]         at dotty.tools.dotc.core.Types$AndType$.balanced(Types.scala:3484)
[error]         at dotty.tools.dotc.core.TypeComparer.andType$$anonfun$1(TypeComparer.scala:2541)
[error]         at dotty.tools.dotc.core.TypeComparer.liftIfHK(TypeComparer.scala:2579)
[error]         at dotty.tools.dotc.core.TypeComparer.andTypeGen(TypeComparer.scala:2519)
[error]         at dotty.tools.dotc.core.TypeComparer.andType(TypeComparer.scala:2541)
[error]         at dotty.tools.dotc.core.TypeComparer.glb(TypeComparer.scala:2354)
[error]         at dotty.tools.dotc.core.TypeComparer.lubArgs(TypeComparer.scala:2413)
[error]         at dotty.tools.dotc.core.TypeComparer$.lubArgs(TypeComparer.scala:3169)
[error]         at dotty.tools.dotc.core.TypeOps$.mergeRefinedOrApplied$1(TypeOps.scala:272)
[error]         at dotty.tools.dotc.core.TypeOps$.orDominator(TypeOps.scala:393)
[error]         at dotty.tools.dotc.core.Types$OrType.join(Types.scala:3546)
[error]         at dotty.tools.dotc.core.Types$Type.goOr$1(Types.scala:916)
[error]         at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:789)
[error]         at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:932)
[error]         at dotty.tools.dotc.core.Types$Type.memberBasedOnFlags(Types.scala:726)
[error]         at dotty.tools.dotc.core.Types$Type.nonPrivateMember(Types.scala:716)
[error]         at dotty.tools.dotc.core.Types$NamedType.memberDenot(Types.scala:2508)
[error]         at dotty.tools.dotc.core.Types$NamedType.reload$1(Types.scala:2841)
[error]         at dotty.tools.dotc.core.Types$NamedType.withPrefix(Types.scala:2856)
[error]         at dotty.tools.dotc.core.Types$NamedType.derivedSelect(Types.scala:2789)
[error]         at dotty.tools.dotc.ast.tpd$TypedTreeCopier.Select(tpd.scala:640)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:241)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:448)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:291)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:450)
[error]         at dotty.tools.dotc.transform.MegaPhase.mapValDef$1(MegaPhase.scala:247)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:252)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:448)
[error]         at dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:461)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:461)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:372)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:450)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:268)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:448)
[error]         at dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:461)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:461)
[error]         at dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:392)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:395)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:450)
[error]         at dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:477)
[error]         at dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:489)
[error]         at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:354)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.immutable.List.foreach(List.scala:333)
[error]         at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:360)
[error]         at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:315)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error]         at dotty.tools.dotc.Run.runPhases$1(Run.scala:337)
[error]         at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:348)
[error]         at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:357)
[error]         at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:71)
[error]         at dotty.tools.dotc.Run.compileUnits(Run.scala:357)
[error]         at dotty.tools.dotc.Run.compileSources(Run.scala:261)
[error]         at dotty.tools.dotc.Run.compile(Run.scala:246)
[error]         at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error]         at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
[error]         at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error]         at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error]         at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
[error]         at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:180)
[error]         at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:178)
[error]         at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:464)
[error]         at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error]         at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error]         at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error]         at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error]         at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:419)
[error]         at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
[error]         at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:406)
[error]         at sbt.internal.inc.Incremental$.apply(Incremental.scala:172)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error]         at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2371)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2321)
[error]         at sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:31)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:47)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:29)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:24)
[error]         at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2319)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:69)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[error]         at sbt.Execute.work(Execute.scala:292)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error]         at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[error]         at java.base/java.lang.Thread.run(Thread.java:1589)
[error] (Compile / compileIncremental) java.lang.AssertionError: assertion failed: ExprType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Int)) in in intersection () ?=> Any & => Int
```
</details>

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -2508,6 +2508,19 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
         NoType
     }
 
+  /** There's a window of vulnerability between ElimByName and Erasure where some
+   *  ExprTypes `=> T` that appear as parameters of function types are not yet converted
+   *  to by-name functions `() ?=> T`. These would cause an assertion violation when
+   *  used as operands of & or |. We fix this on the fly here. As explained in
+   *  ElimByName, we can't fix it beforehand by mapping all occurrences of `=> T` to
+   *  `() ?=> T` since that could lead to cycles.
+   */
+  private def dropExpr(tp: Type): Type = tp match
+    case ExprType(rt) if Phases.elimByNamePhase <= ctx.phase && !ctx.erasedTypes =>
+      defn.ByNameFunction(rt)
+    case _ =>
+      tp
+
   private def andTypeGen(tp1: Type, tp2: Type, op: (Type, Type) => Type,
       original: (Type, Type) => Type = _ & _, isErased: Boolean = ctx.erasedTypes): Type = trace(s"andTypeGen(${tp1.show}, ${tp2.show})", subtyping, show = true) {
     val t1 = distributeAnd(tp1, tp2)
@@ -2516,7 +2529,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
       val t2 = distributeAnd(tp2, tp1)
       if (t2.exists) t2
       else if (isErased) erasedGlb(tp1, tp2)
-      else liftIfHK(tp1, tp2, op, original, _ | _)
+      else liftIfHK(dropExpr(tp1), dropExpr(tp2), op, original, _ | _)
         // The ` | ` on variances is needed since variances are associated with bounds
         // not lambdas. Example:
         //
@@ -2561,7 +2574,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
       val t2 = distributeOr(tp2, tp1, isSoft)
       if (t2.exists) t2
       else if (isErased) erasedLub(tp1, tp2)
-      else liftIfHK(tp1, tp2, OrType.balanced(_, _, soft = isSoft), _ | _, _ & _)
+      else liftIfHK(dropExpr(tp1), dropExpr(tp2), OrType.balanced(_, _, soft = isSoft), _ | _, _ & _)
     }
   }
 

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -2318,40 +2318,35 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
     else if !tp2.exists || (tp2 eq WildcardType) then tp1
     else if tp1.isAny && !tp2.isLambdaSub || tp1.isAnyKind || isBottom(tp2) then tp2
     else if tp2.isAny && !tp1.isLambdaSub || tp2.isAnyKind || isBottom(tp1) then tp1
-    else tp2 match
-      case tp2: LazyRef =>
-        glb(tp1, tp2.ref)
-      case _ =>
-        tp1 match
-          case tp1: LazyRef =>
-            glb(tp1.ref, tp2)
-          case _ =>
-            val tp1a = dropIfSuper(tp1, tp2)
-            if tp1a ne tp1 then glb(tp1a, tp2)
-            else
-              val tp2a = dropIfSuper(tp2, tp1)
-              if tp2a ne tp2 then glb(tp1, tp2a)
-              else tp2 match // normalize to disjunctive normal form if possible.
-                case tp2 @ OrType(tp21, tp22) =>
-                  lub(tp1 & tp21, tp1 & tp22, isSoft = tp2.isSoft)
-                case _ =>
-                  tp1 match
-                    case tp1 @ OrType(tp11, tp12) =>
-                      lub(tp11 & tp2, tp12 & tp2, isSoft = tp1.isSoft)
-                    case tp1: ConstantType =>
-                      tp2 match
-                        case tp2: ConstantType =>
-                          // Make use of the fact that the intersection of two constant types
-                          // types which are not subtypes of each other is known to be empty.
-                          // Note: The same does not apply to singleton types in general.
-                          // E.g. we could have a pattern match against `x.type & y.type`
-                          // which might succeed if `x` and `y` happen to be the same ref
-                          // at run time. It would not work to replace that with `Nothing`.
-                          // However, maybe we can still apply the replacement to
-                          // types which are not explicitly written.
-                          NothingType
-                        case _ => andType(tp1, tp2)
+    else
+      def mergedGlb(tp1: Type, tp2: Type): Type =
+        val tp1a = dropIfSuper(tp1, tp2)
+        if tp1a ne tp1 then glb(tp1a, tp2)
+        else
+          val tp2a = dropIfSuper(tp2, tp1)
+          if tp2a ne tp2 then glb(tp1, tp2a)
+          else tp2 match // normalize to disjunctive normal form if possible.
+            case tp2 @ OrType(tp21, tp22) =>
+              lub(tp1 & tp21, tp1 & tp22, isSoft = tp2.isSoft)
+            case _ =>
+              tp1 match
+                case tp1 @ OrType(tp11, tp12) =>
+                  lub(tp11 & tp2, tp12 & tp2, isSoft = tp1.isSoft)
+                case tp1: ConstantType =>
+                  tp2 match
+                    case tp2: ConstantType =>
+                      // Make use of the fact that the intersection of two constant types
+                      // types which are not subtypes of each other is known to be empty.
+                      // Note: The same does not apply to singleton types in general.
+                      // E.g. we could have a pattern match against `x.type & y.type`
+                      // which might succeed if `x` and `y` happen to be the same ref
+                      // at run time. It would not work to replace that with `Nothing`.
+                      // However, maybe we can still apply the replacement to
+                      // types which are not explicitly written.
+                      NothingType
                     case _ => andType(tp1, tp2)
+                case _ => andType(tp1, tp2)
+      mergedGlb(dropExpr(tp1.stripLazyRef), dropExpr(tp2.stripLazyRef))
   }
 
   def widenInUnions(using Context): Boolean =
@@ -2390,7 +2385,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
         if ((tp1 ne tp1w) || (tp2 ne tp2w)) lub(tp1w, tp2w, canConstrain = canConstrain, isSoft = isSoft)
         else orType(tp1w, tp2w, isSoft = isSoft) // no need to check subtypes again
       }
-      mergedLub(tp1.stripLazyRef, tp2.stripLazyRef)
+      mergedLub(dropExpr(tp1.stripLazyRef), dropExpr(tp2.stripLazyRef))
   }
 
   /** Try to produce joint arguments for a lub `A[T_1, ..., T_n] | A[T_1', ..., T_n']` using
@@ -2511,12 +2506,12 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
   /** There's a window of vulnerability between ElimByName and Erasure where some
    *  ExprTypes `=> T` that appear as parameters of function types are not yet converted
    *  to by-name functions `() ?=> T`. These would cause an assertion violation when
-   *  used as operands of & or |. We fix this on the fly here. As explained in
+   *  used as operands of glb or lub. We fix this on the fly here. As explained in
    *  ElimByName, we can't fix it beforehand by mapping all occurrences of `=> T` to
    *  `() ?=> T` since that could lead to cycles.
    */
   private def dropExpr(tp: Type): Type = tp match
-    case ExprType(rt) if Phases.elimByNamePhase <= ctx.phase && !ctx.erasedTypes =>
+    case ExprType(rt) if (Phases.elimByNamePhase <= ctx.phase) && !ctx.erasedTypes =>
       defn.ByNameFunction(rt)
     case _ =>
       tp
@@ -2529,7 +2524,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
       val t2 = distributeAnd(tp2, tp1)
       if (t2.exists) t2
       else if (isErased) erasedGlb(tp1, tp2)
-      else liftIfHK(dropExpr(tp1), dropExpr(tp2), op, original, _ | _)
+      else liftIfHK(tp1, tp2, op, original, _ | _)
         // The ` | ` on variances is needed since variances are associated with bounds
         // not lambdas. Example:
         //
@@ -2574,7 +2569,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
       val t2 = distributeOr(tp2, tp1, isSoft)
       if (t2.exists) t2
       else if (isErased) erasedLub(tp1, tp2)
-      else liftIfHK(dropExpr(tp1), dropExpr(tp2), OrType.balanced(_, _, soft = isSoft), _ | _, _ & _)
+      else liftIfHK(tp1, tp2, OrType.balanced(_, _, soft = isSoft), _ | _, _ & _)
     }
   }
 

--- a/tests/pos/i19548.scala
+++ b/tests/pos/i19548.scala
@@ -1,0 +1,2 @@
+def byName[T](p: => T): T = p
+val test = (if ??? then byName else (??? : ((=> Int) => Int))) (42)

